pkg/storage/redis/redis.go

```diff
@@ -0,0 +1,176 @@
+package redis
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/gomodule/redigo/redis"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	internal_databroker "github.com/pomerium/pomerium/internal/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/storage"
+)
+
+var _ storage.Backend = (*DB)(nil)
+
+// DB wraps redis conn to interact with redis server.
+type DB struct {
+	pool                   *redis.Pool
+	deletePermanentlyAfter int64
+	recordType             string
+	lastVersion            uint64
```

Comment:
Maybe we should set `lastVersion` somehow? As written I believe if you updated a record, then restarted the process, then updated the same record again you'd end up with the same version number (0).

Reply:
@calebdoxsey I'm a bit confused between this `db.lastVersion` and `databroker.Server.version`, would you mind elaborating?

Also in current implementation, I see `db.lastVersion` is increased everytime we insert or update new record to DB, is it intentional?

Reply:
There are two versions: the record version and the server version. If the server version changes the consumer assumes all its current data is invalidated.

The current implementation is not persistent so the server version will change any time the process is restarted. If it were persistent the server version should not change, but the record version needs to increment properly.

I guess since we've decided not to implement actual persistence of the server version, the issue I'm pointing to won't happen.